added zero shot image classification

dulayjm opened the pull request:
👋 Hi, first-time potential contributor here! I noticed no method for `zero-shot-image-classification` in the inference api, and I wrote an implementation to call for that task. 

🤗 Solution: 

* Add post response in `_client.py` to call a zero-shot image classification model (such as CLIP). 
* Added tests and relevant documentation.

TODO:

Test other models

Wauplin commented on src/huggingface_hub/inference/_client.py:
````diff
@@ -773,6 +773,40 @@ def text_to_speech(self, text: str, *, model: Optional[str] = None) -> bytes:
+        >>> score = client.zero_shot_image_classification("https://upload.wikimedia.org/wikipedia/commons/thumb/4/43/Cute_dog.jpg/320px-Cute_dog.jpg", "dog,cat,horse", model="openai/clip-vit-base-patch32")
+        ```
+        """
+        response = self.post(data={image, text}, model=model, task="zero-shot-image-classification")
````
Looking at the payload sent when trying the widget on a model (see screenshot) and the similar [zero-classification-task](https://huggingface.co/docs/api-inference/detailed_parameters#zeroshot-classification-task) docs, the input format is not correct. It is expected to get a base64-encoded image + some parameters.

```suggestion
        response = self.post(
            json={"image": _b64_encode(image), "parameters": {"candidate_labels": text}},
            model=model,
            task="zero-shot-image-classification",
        )
```

![2023-06-26_14-11](https://github.com/huggingface/huggingface_hub/assets/11801849/57d709b2-7ce0-47e5-adb9-f99a54f4d6a1)

Wauplin commented on tests/cassettes/InferenceClientVCRTest.test_zero_image_classification.yaml:
```diff
@@ -0,0 +1,43 @@
+interactions:
```
This file must be generated automatically (here the format is not correct). To do so, you must have `pytest-vcr` installed + run `pytest tests/test_inference_client.py -k "zero" --vcr-record=all` to record the HTTP call. Don't worry if it doesn't work for you, I can take of it :) Just be careful of not committing your HF token in the yaml file.

dulayjm replied:
This was extremely helpful! I could not find the command to generate this properly. With the typing changes, it passes all tests!

Wauplin commented on tests/test_inference_client.py:
```diff
@@ -197,6 +198,10 @@ def test_text_to_speech(self) -> None:
         self.assertIsInstance(audio, bytes)
 
+    def test_zero_shot_image_classification(self) -> None:
+        word = self.client.zero_shot_image_classification(self.image_file, "tree")
+        self.assertIsInstance(word, str)
```
Expected output is not a string. I think you should try with 

```py
output = self.client.zero_shot_image_classification(self.image_file, "tree,woman,cat")
```
and then test the output that should be a list of dicts (see `test_image_classification`).

Wauplin commented on src/huggingface_hub/inference/_client.py:
```diff
@@ -773,6 +773,40 @@ def text_to_speech(self, text: str, *, model: Optional[str] = None) -> bytes:
         response = self.post(json={"inputs": text}, model=model, task="text-to-speech")
         return response.content
 
+    def zero_shot_image_classification(self, image: ContentT, text: Any, *, model: Optional[str] = None) -> str:
```
Also I realized the server doesn't work if your send only 1 single class (e.g. `"tree"`). You must have at least 2 values. I would add a check on the input and raise a `ValueError` if it's not the case. It always best to fail early and avoid sending a failing request to the server.

dulayjm replied:
I agree the stronger typing makes sense here. Added to change

Wauplin reviewed (commented):
Hi @dulayjm! Thanks for opening a PR for this task :tada: Not all tasks are indeed implemented as the client is quite new and we want other tasks to be more community-lead so I'm glad to see your contribution :hugs: 

I reviewed your PR and made some comments to it. I hope everything's clear but don't hesitate to ping me if not. In general, it would be good to review the inputs/output of the method and adapt the corresponding docs/tests :)

dulayjm commented:
@Wauplin Thank you for the suggested changes!! These were very helpful! In response, I implemented the following: 

* Addressed a typo in the docs 
* Change the input type to `List[str]` to enforce a precise expected input type for labels
* Added a raise to `ValueError` if the length of label responses is less than 2. 
* Changed the output response type to reflect the OpenCLIP model `List[ClassificationOutput]`
* Created the VCR testing file
* Updated a better test in `test_inference_client.py` to address the typing and expected outputs for the model and task.

Wauplin commented on src/huggingface_hub/inference/_client.py:
```diff
@@ -773,6 +773,51 @@ def text_to_speech(self, text: str, *, model: Optional[str] = None) -> bytes:
+                Inference Endpoint. This parameter overrides the model defined at the instance level. Defaults to None.
+
+        Returns:
+            `List[Dict]`: List of classification outputs containing the predicted labels and their confidence
```
```suggestion
            `List[Dict]`: List of classification outputs containing the predicted labels and their confidence.
```

Wauplin commented on src/huggingface_hub/inference/_client.py:
````diff
@@ -773,6 +773,51 @@ def text_to_speech(self, text: str, *, model: Optional[str] = None) -> bytes:
+
+        Example:
+        ```py
+        >>> from pathlib import Path
````
```suggestion
```
No need for pathlib

Wauplin commented on src/huggingface_hub/inference/_client.py:
```diff
@@ -773,6 +773,51 @@ def text_to_speech(self, text: str, *, model: Optional[str] = None) -> bytes:
+            raise ValueError("You must specify at least 2 classes to compare. Please specify more than 1 class.")
+
+        response = self.post(
+            json={"image": _b64_encode(image), "parameters": {"candidate_labels": labels}},
```
```suggestion
            json={"image": _b64_encode(image), "parameters": {"candidate_labels": ",".join(labels)}},
```
Since labels is now a `List[str]` it must be converted to a comma-separated string

Wauplin commented on tests/test_inference_client.py:
```diff
@@ -197,6 +198,14 @@ def test_text_to_speech(self) -> None:
         self.assertIsInstance(audio, bytes)
 
+    def test_zero_shot_image_classification(self) -> None:
+        output = self.client.zero_shot_image_classification(self.image_file, "tree,woman,cat")
```
```suggestion
        output = self.client.zero_shot_image_classification(self.image_file, ["tree", "woman", "cat"])
```
must be a list of labels

Wauplin commented on src/huggingface_hub/inference/_client.py:
```diff
@@ -773,6 +773,51 @@ def text_to_speech(self, text: str, *, model: Optional[str] = None) -> bytes:
+        >>> from huggingface_hub import InferenceClient
+        >>> client = InferenceClient()
+
+        >>> client.zero_shot_image_classification("https://upload.wikimedia.org/wikipedia/commons/thumb/4/43/Cute_dog.jpg/320px-Cute_dog.jpg", "dog,cat,horse", model="openai/clip-vit-base-patch32")
```
```suggestion
        >>> client.zero_shot_image_classification("https://upload.wikimedia.org/wikipedia/commons/thumb/4/43/Cute_dog.jpg/320px-Cute_dog.jpg", ["dog", "cat", "horse"], model="openai/clip-vit-base-patch32")
```

we expect a list of strings now

Wauplin reviewed (commented):
Really nice! I made a quick review of the code and left a few minor comments. I'll run some tests and let you know for a final go!

Wauplin reviewed (approved):
Perfect! Great job here @dulayjm :clap:  I'll merge it as soon as the CI is green :green_circle: 

Would you mind sharing for what project you're interested in using this client? I'd be glad to know more :)

dulayjm commented:
@Wauplin Thanks for the review! Happy to see this get merged for HF 🤗 

I'm using this for a project within my company to quickly deploy CLIP (and a variant of it, [fashion-clip](https://huggingface.co/patrickjohncyh/fashion-clip) to quickly and accurately label items of clothing that our users at Capsule (coming soon) can upload to the platform!

Wauplin commented:
Nice!
